plugins/artifacts/artifacts.go

```diff
@@ -0,0 +1,128 @@
+//go:generate go-composite-schema --unexported artifacts payload-schema.yml generated_payloadschema.go
+
+// Package artifacts is responsible for uploading artifacts after builds
+package artifacts
+
+import (
+	"mime"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+func (pluginProvider) NewPlugin(extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{}, nil
+}
+
+type plugin struct {
+	plugins.PluginBase
+}
+
+func (plugin) PayloadSchema() (runtime.CompositeSchema, error) {
+	return payloadSchema, nil
+}
+
+func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	if options.Payload == nil {
+		return plugins.TaskPluginBase{}, nil
+	}
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		payload:        *(options.Payload.(*payload)),
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context *runtime.TaskContext
+	payload payload
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	tp.context = context
+	return nil
+}
+
+func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
+	var err error
+	for _, artifact := range tp.payload {
+		// If expires is set to this time it's either the default value or has been set to an invalid time anyway
+		if time.Time(artifact.Expires).IsZero() {
+			artifact.Expires = tp.context.TaskInfo.Expires
+		}
+		switch artifact.Type {
+		case "directory":
```

Comment:
> Perhaps we should update the interface definition to allow more non-fatal errors?
What errors do you propose, and how do you propose we handle them?

Yeah, maybe we should allow: `InternalError`:
https://github.com/taskcluster/taskcluster-worker/blob/master/engines/errors.go#L117

With the assumption that this must **only** be used for errors that:
 A) Cannot be trigger by the task payload, or behavior of the code running inside the sandbox,
 B) We are confident will not affect the next task the worker runs

Most internal errors like out-of-disk-space, out-of-memory, cannot contact docker socket, container suddenly disappeared, file that I know I created don't exist anymore, missing hardware device, file system is corrupted, docker daemon died, do **not satisfy condition B**.

In cases like above, the sane thing is to report the task exception worker-shutdown, and then polity crash, probably terminate the instance.

In cases like: can't reach the internet, hostname doesn't resolve, credentials are expired or don't work, etc.
The worker _may_ be able to report the task exception internal-error (have the task retried, we probably need to change queue semantics for that), and then try with the next task.

However, if we see a lot of these can't reach the internet, can't resolve DNS, can't upload to S3, credentials expired unexpectedly, etc... Then we should very much die..

@imbstack, We don't want tasks to retry too much, and we definitely don't want a corrupted worker to just churn through the queue, claim tasks and the report them exception... So we have to be extremely careful about what is a non-fatal internal error. Note: errors that potentially leaves something in a dirty state cannot be non-fatal.


If we want to pursue this, rather than just saying everything is fatal. Then we should probably rename `InternalError ` to `NonFatalInternalError` or `CleanInternalError`, suggestions?
Then specify in doc string what it can be used for, and what it absolutely cannot be used for. Before we expand what the interfaces can return...

This could be a follow-up, please file a bug for those... :)